### PR TITLE
Deprecates `hashValues` and `hashList`

### PR DESCRIPTION
**Pull request by werainkhatri**
Deprecates `hashValues` and `hashList` in favour of `Object.hash` and `Object.hashAll` respectively.

Fixes flutter/flutter#85431

Supposed to land after below listed PRs
- [x] https://github.com/flutter/flutter/pull/96109
- [x] https://github.com/flutter/flutter/pull/99505
- [x] https://github.com/flutter/plugins/pull/4994
- [x] https://github.com/flutter/packages/pull/1581
- [x] https://github.com/flutter/samples/pull/1133

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/hash_codes.dart [hunk 1]
```diff
@@ -2,122 +2,33 @@
...
-                              result = _Jenkins.combine(result, arg16);
-                              if (!identical(arg17, _hashEnd)) {
-                                result = _Jenkins.combine(result, arg17);
-                                if (!identical(arg18, _hashEnd)) {
-                                  result = _Jenkins.combine(result, arg18);
-                                  if (!identical(arg19, _hashEnd)) {
-                                    result = _Jenkins.combine(result, arg19);
-                                    if (!identical(arg20, _hashEnd)) {
-                                      result = _Jenkins.combine(result, arg20);
-                                      // I can see my house from here!
```
Comment (Hixie) [inline lib/ui/hash_codes.dart:91]: i'm sad that my joke is going away :-(

File: lib/ui/hash_codes.dart [hunk 2]
```diff
@@ -2,122 +2,33 @@
...
-int hashList(Iterable<Object?>? arguments) {
-  int result = 0;
-  if (arguments != null) {
-    for (final Object? argument in arguments)
-      result = _Jenkins.combine(result, argument);
-  }
-  return _Jenkins.finish(result);
-}
+@Deprecated(
+  'Use Object.hashAll() or Object.hashAllUnordered() instead '
```
Comment (Hixie) [inline lib/ui/hash_codes.dart:31]: missing period on message

File: lib/ui/hash_codes.dart [hunk 3]
```diff
@@ -2,122 +2,33 @@
...
-/// [Iterable], use [hashList]. The output of [hashList] can be used as one of
-/// the arguments to this function.
-///
-/// For example:
-///
-/// ```dart
-/// int hashCode => hashValues(foo, bar, hashList(quux), baz);
-/// ```
+@Deprecated(
+  'Use Object.hash() instead '
```
Comment (Hixie) [inline lib/ui/hash_codes.dart:9]: missing period on message

File: lib/ui/hash_codes.dart [hunk 4]
```diff
@@ -2,122 +2,33 @@
...
-    for (final Object? argument in arguments)
-      result = _Jenkins.combine(result, argument);
-  }
-  return _Jenkins.finish(result);
-}
+@Deprecated(
+  'Use Object.hashAll() or Object.hashAllUnordered() instead '
+  'This feature was deprecated in v2.9.0-0.1.pre'
+)
+int hashList(Iterable<Object?>? arguments) => Object.hashAll(arguments ?? <Object?>[]);
```
Comment (Hixie) [inline lib/ui/hash_codes.dart:34]: can that list be const?

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -3121,7 +3121,7 @@ class ColorFilter implements ImageFilter {
   }
 
   @override
-  int get hashCode => hashValues(_color, _blendMode, hashList(_matrix), _type);
+  int get hashCode => Object.hash(_color, _blendMode, Object.hashAll(_matrix ?? <Object?>[]), _type);
```
Comment (Hixie) [inline lib/ui/painting.dart:3124]: can that list be const?

File: lib/web_ui/lib/src/ui/hash_codes.dart [hunk 6]
```diff
@@ -4,124 +4,30 @@
...
-  static int finish(int hash) {
-    hash = 0x1fffffff & (hash + ((0x03ffffff & hash) << 3));
-    hash = hash ^ (hash >> 11);
-    hash = 0x1fffffff & (hash + ((0x00003fff & hash) << 15));
-    return hash;
-  }
-}
-
+@Deprecated(
+  'Use Object.hash() instead '
```
Comment (Hixie) [inline lib/web_ui/lib/src/ui/hash_codes.dart:8]: period

File: lib/web_ui/lib/src/ui/hash_codes.dart [hunk 7]
```diff
@@ -4,124 +4,30 @@
...
+  Object? arg19, Object? arg20, ]
+  ) => Object.hash(
+    arg01, arg02, arg03, arg04, arg05, arg06,
+    arg07, arg08, arg09, arg10, arg11, arg12,
+    arg13, arg14, arg15, arg16, arg17, arg18,
+    arg19, arg20,
+  );
+
+@Deprecated(
+  'Use Object.hashAll() or Object.hashAllUnordered() instead '
```
Comment (Hixie) [inline lib/web_ui/lib/src/ui/hash_codes.dart:30]: period

File: lib/web_ui/test/hash_codes_test.dart [hunk 8]
```diff
@@ -1,127 +0,0 @@
-// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (Hixie) [inline lib/web_ui/test/hash_codes_test.dart:1]: I would keep this file (with suitable `//ignore`s for the deprecation warnings and with the numbers adjusted as necessary). That will act as an effective test that the deprecated API still works.

File: lib/ui/hash_codes.dart (hunk 1 above)
Comment (werainkhatri) [reply lib/ui/hash_codes.dart:91]: Sorry to take that away 😢 . I loved it btw! Made me laugh IRL.

File: lib/web_ui/test/hash_codes_test.dart (hunk 8 above)
Comment (werainkhatri) [reply lib/web_ui/test/hash_codes_test.dart:1]: That's the thing, the new API (`Object.hash`)'s seed value changes when the system is rerun, so the same input will give different hash values at every restart (so to speak). So I can't set a constant adjusted value. You can test this in dartpad,dev; Run the below code, see the output, reload the page, run it again, the output will be a different number.

```dart
void main() {
  print(Object.hash(0.0, 0.0.));
}
```

I could maybe change the test in a way that computes the hash twice and tests it if they are the same, if that helps.

File: lib/ui/hash_codes.dart [hunk 9]
```diff
@@ -2,122 +2,33 @@
...
+  Object? arg03, Object? arg04,
+  Object? arg05, Object? arg06,
+  Object? arg07, Object? arg08,
+  Object? arg09, Object? arg10,
+  Object? arg11, Object? arg12,
+  Object? arg13, Object? arg14,
+  Object? arg15, Object? arg16,
+  Object? arg17, Object? arg18,
+  Object? arg19, Object? arg20, ]
+  ) => Object.hash(
```
Comment (rakudrama) [inline lib/ui/hash_codes.dart:23]: This forwarding will not work as expected.
Consider
```dart
hashValues(1, 2)
```
The original `hashValues` will call `1.hashCode` and `2.hashCode` do some combining operations.

The new version will call
```dart
Object.hash(1, 2, null, null, null, ... ,null)
```
`Object.hash` will 'see' that it is passed 20 values and call `.hashCode` on all of them and combine them.
This obviously will be slower (20 vs 2 calls to `.hashCode`).

The old `hashValues` and `Object.hash` both use the same trick - a private default sentinel - to detect the number of arguments. Since you can't access the private sentinel used by `Object.hash`, you can't forward arguments in a way that makes it behave 'the same'.

Recommendation:
Change _nothing_ in this file except for adding `@Deprecated` annotations.
If you don't change the code, you can't accidentally make some app worse.

File: lib/ui/hash_codes.dart [hunk 10]
```diff
@@ -2,122 +2,33 @@
...
 
-class _HashEnd { const _HashEnd(); }
-const _HashEnd _hashEnd = _HashEnd();
-
-/// Jenkins hash function, optimized for small integers.
-//
-// Borrowed from the dart sdk: sdk/lib/math/jenkins_smi_hash.dart.
-class _Jenkins {
-  static int combine(int hash, Object? o) {
-    assert(o is! Iterable);
```
Comment (rakudrama) [inline lib/ui/hash_codes.dart:17]: How important is this assertion?

`Object.hash` has no similar assertion, and never will, since t is possible to have an `Iterable` that does define a well-behaved `hashCode`/`==`, e.g. `BuiltList` https://pub.dev/documentation/built_collection/latest/built_collection/BuiltList/hashCode.html)

**Review by rakudrama — COMMENTED** [review]
It would be worth validating performance after this change.

- Are there any regressions in code that predominantly uses the changes to `Object.hash`
- Are there any regressions in unmigrated code that still uses `hashValues`?

File: lib/ui/hash_codes.dart (hunk 9 above)
Comment (kevmoo) [reply lib/ui/hash_codes.dart:23]: Agree with this!

Comment (werainkhatri) [reply lib/ui/hash_codes.dart:23]: Thank you for the review! Makes perfect sense.

Looking back, having little deprecation knowledge, I just took inspiration from a few deprecated members and noticed that they were proxied to their replacements. Should have dug a little deeper.

Nevertheless, will fix this ASAP.

Comment (kevmoo) [reply lib/ui/hash_codes.dart:23]: If you *can* cleanly move to the new thing, it's generally a good thing. This is a gnarly case, though.

File: lib/web_ui/test/hash_codes_test.dart (hunk 8 above)
Comment (werainkhatri) [reply lib/web_ui/test/hash_codes_test.dart:1]: cc @yjbanov

File: lib/ui/hash_codes.dart (hunk 10 above)
Comment (werainkhatri) [reply lib/ui/hash_codes.dart:17]: cc @Hixie @sigurdm

File: lib/ui/hash_codes.dart (hunk 9 above)
Comment (werainkhatri) [reply lib/ui/hash_codes.dart]: Done!

File: lib/web_ui/test/hash_codes_test.dart (hunk 8 above)
Comment (werainkhatri) [reply lib/web_ui/test/hash_codes_test.dart]: Restored the test.

File: lib/ui/hash_codes.dart (hunk 1 above)
Comment (werainkhatri) [reply lib/ui/hash_codes.dart]: Your joke lives on 🖖🏾

File: lib/web_ui/lib/src/ui/hash_codes.dart [hunk 11]
```diff
@@ -4,96 +4,91 @@
...
+/// For example:
+///
+/// ```dart
+/// int hashCode => hashValues(foo, bar, hashList(quux), baz);
+/// ```
+@Deprecated(
+  'Use Object.hash() instead. '
+  'This feature was deprecated in v2.9.0-0.1.pre'
+)
 int hashValues(
```
Comment (sigurdm) [inline lib/web_ui/lib/src/ui/hash_codes.dart:46]: Drive-by comment:
You could possibly just redirect to Object.hash here instead of keeping the implementation.

Same below

File: lib/ui/hash_codes.dart (hunk 10 above)
Comment (sigurdm) [reply lib/ui/hash_codes.dart]: I don't think it is very important - I think it is guarding against thinking you are doing a deep hash while you most probably are not (the default List has identity-based hashCode). This is not foolproof, and for example won't guard against Maps that have the same issue.

I agree that if an `Iterable` implements a non-identity hashCode then the assertion is not-needed.

Not sure if there is a way of asking that question.

File: lib/web_ui/lib/src/ui/hash_codes.dart (hunk 11 above)
Comment (werainkhatri) [reply lib/web_ui/lib/src/ui/hash_codes.dart:46]: I tried passing in all the args from `hashValues` to `Object.hash`, but as pointed out in the above thread (unresolved it now), it will provide inconsistent hash values. If you are implying a different way to redirect, could you elaborate?

Comment (sigurdm) [reply lib/web_ui/lib/src/ui/hash_codes.dart:46]: Ah sorry - I had not seen that discussion - no I did not have anything more clever in mind :)

**Comment by skia-gold** [comment]
Gold has detected about 38 new digest(s) on patchset 5.
View them at https://flutter-engine-gold.skia.org/cl/github/30674

File: lib/web_ui/lib/src/engine/text/ruler.dart [hunk 12]
```diff
@@ -133,16 +133,15 @@ class TextDimensions {
...
   /// The height of the paragraph being measured.
   double get height {
     double cachedHeight = _readAndCacheMetrics().height as double;
     if (browserEngine == BrowserEngine.firefox &&
-      // In the flutter tester environment, we use a predictable-size for font
-      // measurement tests.
-      !ui.debugEmulateFlutterTesterEnvironment) {
+        // In the flutter tester environment, we use a predictable-size for font
+        // measurement tests.
+        !ui.debugEmulateFlutterTesterEnvironment) {
```
Comment (Hixie) [inline lib/web_ui/lib/src/engine/text/ruler.dart:144]: unexpected indent

**Review by Hixie — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/text/ruler.dart (hunk 12 above)
Comment (werainkhatri) [reply lib/web_ui/lib/src/engine/text/ruler.dart]: fixed!

**Comment by werainkhatri** [comment]
waiting for https://github.com/flutter/packages/pull/1581 to be merged, to prevent any rolling deprecation issues :)

**Comment by a-siva** [comment]
This appears to break the framework builds and the Flutter HHH builds

**Comment by jason-simmons** [comment]
Landed https://github.com/flutter/flutter/commit/116d6579c7fb40c67dada6dd9d50e2524d06a40e in the framework, which makes it compatible with this engine PR.
